src/Marten/Events/Daemon/HighWater/HighWaterAgent.cs

```diff
@@ -0,0 +1,144 @@
+{
+    internal class HighWaterAgent : IDisposable
+    {
+        private readonly IHighWaterDetector _detector;
+        private readonly ShardStateTracker _tracker;
+        private readonly ILogger<IProjection> _logger;
+        private readonly DaemonSettings _settings;
+        private readonly CancellationToken _token;
+        private readonly Timer _timer;
+        private Task<Task> _loop;
+
+        private HighWaterStatistics _current;
+
+        // ReSharper disable once ContextualLoggerProblem
+        public HighWaterAgent(IHighWaterDetector detector, ShardStateTracker tracker, ILogger<IProjection> logger, DaemonSettings settings, CancellationToken token)
+        {
+            _detector = detector;
+            _tracker = tracker;
+            _logger = logger;
+            _settings = settings;
+            _token = token;
+
+            _timer = new Timer(_settings.HealthCheckPollingTime.TotalMilliseconds) {AutoReset = true};
+            _timer.Elapsed += TimerOnElapsed;
+        }
+
+        public void Start()
+        {
+
+            _loop = Task.Factory.StartNew(DetectChanges, TaskCreationOptions.LongRunning | TaskCreationOptions.AttachedToParent);
+
+            _timer.Start();
+
+            _logger.LogInformation("Started HighWaterAgent.");
+        }
+
+
+        private async Task DetectChanges()
+        {
+            // TODO -- need to put some retry & exception handling here.
+            try
+            {
+                _current = await _detector.Detect(_token);
+
+                _tracker.MarkHighWater(_current.CurrentMark);
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(e, "Failed while making the initial determination of the high water mark");
+            }
+
+            await Task.Delay(_settings.FastPollingTime, _token);
+
+            while (!_token.IsCancellationRequested)
+            {
+                HighWaterStatistics statistics = null;
+                try
+                {
+                    statistics = await _detector.Detect(_token);
+                }
+                catch (Exception e)
+                {
+                    _logger.LogError("Failed while trying to detect high water statistics", e);
+                    await Task.Delay(_settings.SlowPollingTime, _token);
+                    continue;
+                }
+
+                var status = statistics.InterpretStatus(_current);
+
+                switch (status)
+                {
+                    case HighWaterStatus.Changed:
+                        await markProgress(statistics, _settings.FastPollingTime);
+                        break;
+
+                    case HighWaterStatus.CaughtUp:
+                        await markProgress(statistics, _settings.SlowPollingTime);
+                        break;
+
+                    case HighWaterStatus.Stale:
+                        var safeHarborTime = _current.Timestamp.Add(_settings.LeadingEdgeBuffer);
```

Comment:
Same consideration as above about the timestamp vs numeric offset and time skew as above https://github.com/JasperFx/marten/pull/1712#discussion_r563291852.

Reply:
Same comment. This is to slide around "stale missing sequence" values. We're not using this in any way where we need precision. The "safe harbor" time would be something like "just assume anything older than 5 seconds is good". Easier to talk about this in Zoom.

When the HighWaterDetector "finds" the statistics, it starts looking from the last previously known sequence. Keep that in mind here.

So the case here is:

1. fetch the statistics once, the last good number is 1000, but the sequence is at 1300.
2. fetch the statistics again on the regular polling interval, the last good number is 1000, but the sequence has advanced to 1500. We can tell that there's stale data where transactions are either being *really slow finishing up, but have already reserved the sequence numbers. The current async daemon *sucks* because you always have to be very careful about whether or not missing sequence numbers are real and just in flight, or the result of a failure. The new tombstone event thing *should* make that problem mostly go away, but only mostly. So at some point we have to assume that it's safe to pick at any events older than the "safe harbor time". 
3. wait a calculated moment, and look for the highest sequence number that w/o gaps higher than the "safe harbor" timestamp

At no time are we using that timestamp as the determination for what evens have been published or not, except for the stale sequence gap issue. Which shouldn't happen unless you get funky database connectivity issues. And even there, you're not needing a lot of precision in the timestamp values, so I'm not concerned with the node drift issue.